src/mappings/HistoryElements.ts

```diff
@@ -1,22 +1,108 @@
+import {AnyTuple} from "@polkadot/types/types/codec";
 
 export async function handleHistoryElement(extrinsic: SubstrateExtrinsic): Promise<void> {
     const { isSigned } = extrinsic.extrinsic;
     if (isSigned) {
-        const element = new HistoryElement(extrinsic.extrinsic.hash.toString());
-        element.address = extrinsic.extrinsic.signer.toString()
-        element.timestamp = timestamp(extrinsic.block)
-
-        element.extrinsic = {
-            hash: extrinsic.extrinsic.hash.toString(),
-            module: extrinsic.extrinsic.method.section,
-            call: extrinsic.extrinsic.method.method,
-            success: extrinsic.success,
-            fee: exportFeeFromDepositEvent(extrinsic)?.toString()
+        let failedTransfers = findFailedTransferCalls(extrinsic)
+        if (failedTransfers != null) {
+            await saveFailedTransfers(failedTransfers, extrinsic)
+        } else {
+            await saveExtrinsic(extrinsic)
         }
+    }
+}
+
+async function saveFailedTransfers(transfers: Transfer[], extrinsic: SubstrateExtrinsic): Promise<void> {
+    let promises = transfers.map(transfer => {
+        const elementFrom = new HistoryElement(transfer.extrinsicId+`-from`);
+        elementFrom.address = transfer.from
+        elementFrom.timestamp = timestamp(extrinsic.block)
+        elementFrom.transfer = transfer
+
+        const elementTo = new HistoryElement(transfer.extrinsicId+`-to`);
+        elementTo.address = transfer.to
+        elementTo.timestamp = timestamp(extrinsic.block)
+        elementTo.transfer = transfer
+
+        return [elementTo.save(), elementFrom.save()]
+    })
+    await Promise.allSettled(promises)
+}
 
-        await element.save()
+async function saveExtrinsic(extrinsic: SubstrateExtrinsic): Promise<void> {
+    const element = new HistoryElement(extrinsic.extrinsic.hash.toString());
+    element.address = extrinsic.extrinsic.signer.toString()
+    element.timestamp = timestamp(extrinsic.block)
+    element.extrinsic = {
+        hash: extrinsic.extrinsic.hash.toString(),
+        module: extrinsic.extrinsic.method.section,
+        call: extrinsic.extrinsic.method.method,
+        success: extrinsic.success,
+        fee: exportFeeFromDepositEvent(extrinsic)?.toString()
     }
+    await element.save()
+}
+
+/// Success Transfer emits Transfer event that is handled at Transfers.ts handleTransfer()
+function findFailedTransferCalls(extrinsic: SubstrateExtrinsic): Transfer[] | null {
+    if (extrinsic.success) {
+        return null;
+    }
+
+    let transferCallsArgs = determineTransferCallsArgs(extrinsic.extrinsic.method)
+    if (transferCallsArgs.length == 0) {
+        return null;
+    }
+
+    let sender = extrinsic.extrinsic.signer
+    return transferCallsArgs.map(tuple => {
+        let blockNumber = extrinsic.block.block.header.number.toString()
+        return {
+            amount: tuple[1].toString(),
+            from: sender.toString(),
+            to: tuple[0],
```

Comment:
Interesting thing. It can be a multi address here or raw accountId. If we leave it as it is then it will be a problem to interpret receiver for failed transfer on the client

Reply:
Can we check if there is an option to get accountId

Reply:
Ok, looks like js library handles this